Add fl hub example

holgerroth opened the pull request:
Example of hierarchical interaction between several levels of nvflare FL systems, e.g. Tier-1 (hub) and Tier-2 (sub-systems).
Adds example to show a two-level hub system.

Tested with https://github.com/NVIDIA/NVFlare/pull/1739

holgerroth commented:
/build

yanchengnv commented on nvflare/app_common/hub/hub_persistor.py:
```diff
@@ -0,0 +1,20 @@
+from nvflare.apis.fl_context import FLContext
```
I don't think we need this file any more - it's only used for the V1 solution.

yanchengnv commented on nvflare/app_common/hub/hub_shareable_generator.py:
```diff
@@ -0,0 +1,94 @@
+# Copyright (c) 2021-2022, NVIDIA CORPORATION.  All rights reserved.
```
Don't need this for v2.

yanchengnv commented on nvflare/app_common/resource_managers/combo_manager.py:
```diff
@@ -0,0 +1,106 @@
+# Copyright (c) 2021-2022, NVIDIA CORPORATION.  All rights reserved.
```
No needed. Please remove.

yanchengnv reviewed (changes requested):
Please remove those unnecessary files (they were used for V1 hub solution).
Make sure the cross-site validation still works in both normal and hub cases.

holgerroth commented:
> Please remove those unnecessary files (they were used for V1 hub solution). Make sure the cross-site validation still works in both normal and hub cases.

Removed those files and updated `CrossSiteModelEval` and `ValidationJsonGenerator` to support DXO collections from the 2nd tier systems. The cross-site validation output now looks like so:
```
{
	"t1_client_b": {
		"site_b-1": {
			"accuracy": 6.68965538214698
		},
		"site_b-2": {
			"accuracy": 6.090738110449937
		}
	},
	"t1_client_a": {
		"site_a-1": {
			"accuracy": 6.969653228011714
		}
	}
}
```

holgerroth commented:
/build

holgerroth commented:
/build

holgerroth commented:
/build

yanchengnv commented on examples/advanced/fl_hub/config/site_a/hub_client.json:
```diff
@@ -0,0 +1,28 @@
+        "args": {
+          "pipe_id": "pipe",
+          "task_wait_time": 600,
+          "result_poll_interval": 0.5
```
You can reduce it to 0.1, or just omit it so default will be used here.

yanchengnv commented on examples/advanced/fl_hub/config/site_a/hub_server.json:
```diff
@@ -0,0 +1,23 @@
+      "path": "nvflare.app_common.hub.hub_controller.HubController",
+      "args": {
+        "pipe_id": "pipe",
+        "task_wait_time": 60,
```
You don't have to specify this. If not specified, the HubController will keep waiting until either a task is received or the peer is disconnected (based on heartbeat).

yanchengnv commented on examples/advanced/fl_hub/jobs/numpy-cross-val/meta.json:
```diff
@@ -0,0 +1,11 @@
+  "name": "numpy-cross-val",
+  "resource_spec": {},
+  "min_clients" : 2,
+  "deploy_map": {
```
You no longer have to define separate apps for the two T2 systems, unless you want them to be different.
You can just define one single app, and put "operators" in the config_fed_server.json.

yanchengnv reviewed (commented):
See my comments and suggestions.

holgerroth commented:
> @holgerroth @yanchengnv do we need to update this PR or close this PR/created a new PR ?

This one. I already rebased it on the latest dev branch.

yanchengnv reviewed (approved):
The comments about config are also documented in FL Hub doc.
